prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+				if (ret != 1)
+					FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,
+					       "Invalid signal\n");
+				continue;
+			}
+
+		} else {
+			break;
+		}
+
+		conn_fd = accept(pep->cm.sock, NULL, 0);
+		if (conn_fd < 0) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to accept: %d\n", errno);
+			continue;
+		}
+
+		handle = calloc(1, sizeof(*handle));
+		if (!handle) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"cannot allocate memory\n");
+
+			ofi_close_socket(conn_fd);
+			break;
+		}
+		handle->conn_fd = conn_fd;
+
+		entry_sz = sizeof(*cm_entry);
+		cm_entry = calloc(1, entry_sz);
+		if (!cm_entry) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"cannot allocate memory\n");
+			goto err;
+		}
+
+		cm_entry->fid = &pep->pep.fid;
+		cm_entry->info = &pep->info;
+		cm_entry->info->handle = &handle->handle;
+
+		/* report the conn req to the associated eq */
+		if (fi_eq_write(&pep->eq->eq_fid, FI_CONNREQ, cm_entry, entry_sz, 0) < 0)
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"Error in writing to EQ\n");
+		free(cm_entry);
+	}
+	return NULL;
+err:
+	ofi_close_socket(conn_fd);
+	free(cm_entry);
+	free(handle);
+	return NULL;
+}
+
+static int tcpx_pep_listen(struct fid_pep *pep)
+{
+	struct tcpx_pep *_pep;
+	/* struct fi_info *fi; */
+	struct addrinfo hints, *addr_result, *iter;
+	char sa_ip[INET_ADDRSTRLEN] = {0};
+	char sa_port[NI_MAXSERV] = {0};
+	int ret;
+
+	_pep = container_of(pep,struct tcpx_pep, pep);
+
+	/* fi = &_pep->info; */
+
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_flags = AI_PASSIVE;
+
+	memcpy(sa_ip, inet_ntoa(_pep->src_addr.sin_addr), INET_ADDRSTRLEN);
+	sprintf(sa_port, "%d", ntohs(_pep->src_addr.sin_port));
+
+	ret = getaddrinfo(sa_ip, sa_port, &hints, &addr_result);
+	if (ret) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"no available AF_INET address service:%s, %s\n",
+			sa_port, gai_strerror(ret));
+		return -FI_EINVAL;
+	}
+
+	for (iter = addr_result; iter; iter = iter->ai_next) {
+		_pep->cm.sock = socket(iter->ai_family, iter->ai_socktype,
+				      iter->ai_protocol);
+		if (_pep->cm.sock >= 0) {
+			tcpx_set_sockopts(_pep->cm.sock);
+			if (!bind(_pep->cm.sock, addr_result->ai_addr, addr_result->ai_addrlen))
+				break;
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to bind listener: f%s\n", strerror(errno));
+			ofi_close_socket(_pep->cm.sock);
+			_pep->cm.sock = -1;
+		}
+	}
+
+	freeaddrinfo(addr_result);
+
+	if (_pep->cm.sock < 0) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create listener: %s\n", strerror(errno));
+		return -FI_EIO;
+	}
+
+	if (listen(_pep->cm.sock, TCPX_MAX_SOCK_REQS)) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"socket listen failed\n");
+		goto out;
+	}
+
+	_pep->cm.do_listen = 1;
+
+	if (pthread_create(&_pep->cm.listener_thread, NULL,
+			   tcpx_cm_listener_thread, (void *)_pep)) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create cm thread\n");
+		goto out;
+	}
```

Comment:
You mentioned this, but we will want a single thread per fabric or domain, rather than per PEP.